modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1405,6 +1408,20 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
 		resolve_pending_lambda_bodies();
 	}
 
+	// Resolve base abstract class/method implementation requirements.
+	const GDScriptParser::ClassNode *base_class = p_class->base_type.class_type;
+	while (base_class) {
+		if (p_class->is_abstract || !base_class->is_abstract) {
+			break;
+		}
+		for (GDScriptParser::ClassNode::Member t : base_class->members) {
+			if (t.type == GDScriptParser::ClassNode::Member::Type::FUNCTION && t.function->is_abstract && !p_class->has_function(t.function->identifier->name)) {
+				push_error(vformat(R"(Class%s must implement %s from abstract class %s.)", p_class->identifier == nullptr ? "" : vformat(R"( %s)", p_class->identifier->name), t.function->identifier->name, base_class->identifier == nullptr ? "<main>" : base_class->identifier->name), p_class);
+			}
+		}
+		base_class = base_class->base_type.class_type;
+	}
+
```

Comment:
```suggestion
	if (!p_class->is_abstract) {
		HashSet<StringName> implemented_funcs;
		const GDScriptParser::ClassNode *base_class = p_class;
		while (base_class != nullptr) {
			if (!base_class->is_abstract && base_class != p_class) {
				break;
			}
			for (GDScriptParser::ClassNode::Member member : base_class->members) {
				if (member.type == GDScriptParser::ClassNode::Member::FUNCTION) {
					if (member.function->is_abstract) {
						if (!implemented_funcs.has(member.function->identifier->name)) {
							push_error(vformat(R"*(Class "%s" is not marked as abstract, so it must implement all abstract methods. The first one detected is "%s()".)*",
									p_class->identifier == nullptr ? p_class->fqcn.get_file() : p_class->identifier->name, member.function->identifier->name));
							break;
						}
					} else {
						implemented_funcs.insert(member.function->identifier->name);
					}
				}
			}
			if (base_class->base_type.kind == GDScriptParser::DataType::CLASS) {
				base_class->base_type.class_type;
			} else if (base_class->base_type.kind == GDScriptParser::DataType::SCRIPT) {
				Ref<GDScriptParserRef> parser_ref = get_parser_for(base_class->base_type.script_path);
				ERR_FAIL_COND(parser_ref.is_null());
				base_class = parser_ref->get_parser()->head;
			} else {
				break;
			}
		}
	}
```

I haven't tested this.

Reply:
Using C# as an example, they error out for every method not implemented in a subclass:
![image](https://github.com/godotengine/godot/assets/56272643/1cc31a2f-19e7-4ff9-af5b-8f4d0c3061fc)

Reply:
We need to display different messages depending on where the unimplemented abstract method is found: in the current class or in the base classes:

```c++
if (base_class == p_class) {
    push_error(vformat(R"(Class "%s" contains abstract method(s), so it must have "@abstract" annotation.)",
            p_class->identifier == nullptr ? p_class->fqcn.get_file() : p_class->identifier->name), p_class);
} else {
    push_error(vformat(R"(Class "%s" is not marked as abstract, so it must implement all inherited abstract methods.)",
            p_class->identifier == nullptr ? p_class->fqcn.get_file() : p_class->identifier->name), p_class);
}
```